lib/Serialization/Deserialization.cpp

```diff
@@ -4428,7 +4457,7 @@ Expected<Type> ModuleFile::getTypeChecked(TypeID TID) {
     decls_block::UnboundGenericTypeLayout::readRecord(scratch,
                                                       genericID, parentID);
 
-    auto genericDecl = cast<NominalTypeDecl>(getDecl(genericID));
+    auto genericDecl = cast<GenericTypeDecl>(getDecl(genericID));
```

Comment:
Oh, cute. This is a bug fix. Here's a test case:

```
Slavas-MBP:swift slava$ cat ta.swift
import other

func f(_: GG<Int>) {}
Slavas-MBP:swift slava$ cat other.swift
public struct Outer { public typealias G<T> = T }

public typealias GG = Outer.G
```

Building other.swift by itself without WMO is enough to crash because we deserialize the unbound generic type in merge-modules:

```
Slavas-MBP:swift slava$ ../build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc other.swift -emit-module
Assertion failed: (isa<X>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file /Users/slava/new/llvm/include/llvm/Support/Casting.h, line 236.
0  swift                    0x00000001089da068 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x00000001089da716 SignalHandler(int) + 454
2  libsystem_platform.dylib 0x00007fff8a50abba _sigtramp + 26
3  swift                    0x000000010627149a swift::ModuleFile::getGenericSignatureOrEnvironment(llvm::PointerEmbeddedInt<unsigned int, 31>, bool) + 1450
4  libsystem_c.dylib        0x00007fff8a391420 abort + 129
5  libsystem_c.dylib        0x00007fff8a358893 basename_r + 0
6  swift                    0x0000000106278883 swift::ModuleFile::getTypeChecked(llvm::PointerEmbeddedInt<unsigned int, 31>) + 9411
7  swift                    0x0000000106266130 swift::ModuleFile::getType(llvm::PointerEmbeddedInt<unsigned int, 31>) + 32
8  swift                    0x000000010626a092 swift::ModuleFile::getDeclChecked(llvm::PointerEmbeddedInt<unsigned int, 31>, llvm::Optional<swift::DeclContext*>) + 16066
9  swift                    0x000000010626530e swift::ModuleFile::getDecl(llvm::PointerEmbeddedInt<unsigned int, 31>, llvm::Optional<swift::DeclContext*>) + 78
10 swift                    0x00000001062a1b68 swift::ModuleFile::getTopLevelDecls(llvm::SmallVectorImpl<swift::Decl*>&) + 920
11 swift                    0x000000010668279f swift::ModuleDecl::getTopLevelDecls(llvm::SmallVectorImpl<swift::Decl*>&) const + 63
12 swift                    0x0000000105e6f38e swift::ClassHierarchyAnalysis::init() + 78
13 swift                    0x0000000105e62d39 swift::createClassHierarchyAnalysis(swift::SILModule*) + 153
14 swift                    0x0000000105f2a73c swift::SILPassManager::SILPassManager(swift::SILModule*, llvm::StringRef) + 460
15 swift                    0x0000000105f361a7 swift::runSILDiagnosticPasses(swift::SILModule&) + 135
16 swift                    0x00000001056b5ca8 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 12952
17 swift                    0x0000000105670280 main + 3312
18 libdyld.dylib            0x00007fff8a2fd255 start + 1
```

If you build it with WMO on, other.swift builds, but you'll get the same crash when you build ta.swift.

Do you mind adding this as as test case somewhere?